From 6ecbb870f24f5c5c4a5b548166ac292801adda84 Mon Sep 17 00:00:00 2001
From: Mohammed Naser <mnaser@vexxhost.com>
Date: Sun, 19 Feb 2023 21:39:46 +0000
Subject: [PATCH] [capi] Avoid going through Heat for worker list

By default, Magnum UI goes through Heat to get the list of nodes
which is not correct since it's making an assumption that Heat
is always in use.

The fix for this would be to make sure that Magnum has a list of
all the VMs in it's database (or some sort of API call that
returns them all from the driver) but that's quite a big amount
of work to implement for now.

So for now, if stack_id doesn't look like a UUID, we assume it
is deployed using Clsuter API driver for Magnum and look up with
that alternative method instead.

(cherry picked from commit 6f31cc5cacf23398b76392922ee9863d50aa9e7e)
(cherry picked from commit d44f16f13a89d7fb00d3d949a392d638ce2d0cc8)
(cherry picked from commit 72122e350429590e9002058e7e35c4dcc94d2d4f)

Change-Id: I62bb6352e54d72296a9639856f8dd3de19ceb869
Signed-off-by: ricolin <rlin@vexxhost.com>
---
 magnum_ui/api/rest/magnum.py | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/magnum_ui/api/rest/magnum.py b/magnum_ui/api/rest/magnum.py
index 19fe816..4ae662b 100644
--- a/magnum_ui/api/rest/magnum.py
+++ b/magnum_ui/api/rest/magnum.py
@@ -17,6 +17,8 @@ import re
 
 from collections import defaultdict
 
+from oslo_utils import uuidutils
+
 from django.conf import settings
 from django.http import HttpResponse
 from django.http import HttpResponseNotFound
@@ -229,6 +231,19 @@ class ClusterResize(generic.View):
 
     url_regex = r'container_infra/clusters/(?P<cluster_id>[^/]+)/resize$'
 
+    def _cluster_api_resize_get(self, request, cluster):
+        search_opts = {"name": "%s-" % cluster["stack_id"]}
+        servers = api.nova.server_list(request, search_opts=search_opts)[0]
+
+        worker_nodes = []
+        for server in servers:
+            control_plane_prefix = "%s-control-plane" % cluster["stack_id"]
+            if not server.name.startswith(control_plane_prefix):
+                worker_nodes.append({"name": server.name, "id": server.id})
+
+        return {"cluster": change_to_id(cluster),
+                "worker_nodes": worker_nodes}
+
     @rest_utils.ajax()
     def get(self, request, cluster_id):
         """Get cluster details for resize"""
@@ -238,6 +253,9 @@ class ClusterResize(generic.View):
             print(e)
             return HttpResponseNotFound()
 
+        if not uuidutils.is_uuid_like(cluster["stack_id"]):
+            return self._cluster_api_resize_get(request, cluster)
+
         try:
             ngs = magnum.nodegroup_list(request, cluster_id)
             nodegroups = [n.to_dict() for n in ngs]
-- 
2.25.1

